sound/soc/sof/ipc4.c

```diff
@@ -657,7 +658,35 @@ static const struct sof_ipc_pm_ops ipc4_pm_ops = {
 	.set_core_state = sof_ipc4_set_core_state,
 };
 
+static int sof_ipc4_init(struct snd_sof_dev *sdev)
+{
+	struct sof_ipc4_fw_data *ipc4_data = sdev->private;
+
+	xa_init_flags(&ipc4_data->fw_lib_xa, XA_FLAGS_ALLOC);
+
+	return 0;
+}
+
+static void sof_ipc4_exit(struct snd_sof_dev *sdev)
+{
+	struct sof_ipc4_fw_data *ipc4_data = sdev->private;
+	struct sof_ipc4_fw_library *fw_lib;
+	unsigned long lib_id;
+
+	xa_for_each(&ipc4_data->fw_lib_xa, lib_id, fw_lib) {
+		/* Do not release the basefw library */
```

Comment:
OK, I'll update the comment.

We special case the basefw because it is a special case. The basefw is the kernel for the DSP and it is loaded by the generic loader code. The basefw is the kernel for the DSP, there is no difference on how the kernel is loaded by the SOF stack and it is handled by a generic code.

For IPC4 we just 'borrow' what has been already loaded as Library #0 to be able to handle generic code to walk the libraries later.
This patch do abstracts the difference between basefw and external libraries by representing the already present basefw as a library.
We _could_ move this whole library concept up to the highest level and force it to IPC3, which does not have support for external libraries, afaik.
Moving this up would create another question: is this library and modules inside of a library concept universal? HOw do we abstract away this detail? Do we represent firmware files and then leave the interpretation to IPC code? How the information will flow from IPC to generic and backwards? Is it written in stone that UUID is the name that needs to be looked up by other IPC types? So on.